velox/functions/prestosql/DecimalVectorFunctions.cpp

```diff
@@ -160,28 +160,85 @@ class DecimalBetweenFunction : public exec::VectorFunction {
+          result->asUnchecked<FlatVector<bool>>()->set(
+              row, rawA[row] >= constantB && rawA[row] <= constantC);
+        });
+        return;
+      }
+
+      if (args[1]->isConstantEncoding() && args[2]->isFlatEncoding()) {
+        auto constantB = args[1]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+        auto flatC = args[2]->asUnchecked<FlatVector<A>>();
+        auto rawC = flatC->mutableRawValues();
+        context.applyToSelectedNoThrow(rows, [&](auto row) {
+          result->asUnchecked<FlatVector<bool>>()->set(
+              row, rawA[row] >= constantB && rawA[row] <= rawC[row]);
+        });
+        return;
+      }
+
+      if (args[1]->isConstantEncoding() && args[2]->isFlatEncoding()) {
+        auto flatB = args[1]->asUnchecked<FlatVector<A>>();
+        auto rawB = flatB->mutableRawValues();
+        auto constantC = args[2]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+        context.applyToSelectedNoThrow(rows, [&](auto row) {
+          result->asUnchecked<FlatVector<bool>>()->set(
+              row, rawA[row] >= rawB[row] && rawA[row] <= constantC);
+        });
+        return;
+      }
+
+      if (args[1]->isFlatEncoding() && args[2]->isFlatEncoding()) {
+        auto flatB = args[1]->asUnchecked<FlatVector<A>>();
+        auto rawB = flatB->mutableRawValues();
+        auto flatC = args[2]->asUnchecked<FlatVector<A>>();
+        auto rawC = flatC->mutableRawValues();
+        context.applyToSelectedNoThrow(rows, [&](auto row) {
+          result->asUnchecked<FlatVector<bool>>()->set(
+              row, rawA[row] >= rawB[row] && rawA[row] <= rawC[row]);
+        });
+        return;
+      }
     } else {
-      // Path if first argument is encoded.
+      // Fast path when the first argument is encoded but the second and third
+      // are constants.
       exec::DecodedArgs decodedArgs(rows, args, context);
       auto a = decodedArgs.at(0);
-      context.applyToSelectedNoThrow(rows, [&](auto row) {
-        auto value = a->valueAt<A>(row);
-        result->asUnchecked<FlatVector<bool>>()->set(
-            row, value >= constantB && value <= constantC);
-      });
+
+      if (args[1]->isConstantEncoding() && args[2]->isConstantEncoding()) {
+        auto constantB = args[1]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+        auto constantC = args[2]->asUnchecked<SimpleVector<A>>()->valueAt(0);
+        context.applyToSelectedNoThrow(rows, [&](auto row) {
+          auto value = a->valueAt<A>(row);
+          result->asUnchecked<FlatVector<bool>>()->set(
+              row, value >= constantB && value <= constantC);
+        });
+        return;
+      }
     }
+
+    // Decode the input in all other cases.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto aDecoded = decodedArgs.at(0);
+    auto bDecoded = decodedArgs.at(1);
+    auto cDecoded = decodedArgs.at(2);
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto aValue = aDecoded->valueAt<A>(row);
```

Comment:
I've generally seen code that gets the values buffer using data() for DecodedVector and then uses indices() to index for particular positions in the vector. Did you consider that approach here ?

Reply:
Good question. I did not specifically consider this approach but I have followed a similar code in this file: https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/DecimalVectorFunctions.cpp#L177-L183. 

I double checked the implementation - it appears that `valueAt` is a wrapper for `data()[index]`, here is the actual code: 
```c++
const T* data() const {
  return reinterpret_cast<const T*>(data_);
}

template <typename T>
T valueAt(vector_size_t idx) const {
  return reinterpret_cast<const T*>(data_)[index(idx)];
}
```

It seems either could be used, I just went with `valueAt` for consistency in the file.

Reply:
So obtaining the data() pointer once and then indexing it using index for each row in the loop, avoids the expensive reinterpret_cast that happens for each row in the above code. Please can you try that approach.

Reply:
Thanks, Aditi. Yes, I have updated the code to use data() outside of the loop + index() instead of valueAt().
Also, while making the change I found and fix a bug in the code so many thanks for noticing this!

Reply:
`reinterpret_cast` is a compile-time directive. It has no runtime cost. It is always better to use an API where it is possible to do so.